### ISSUE TITLE
Possible for client to be in a nonexistent company on a server; but then the client crashes.

### ISSUE THREAD

**Issue by DorpsGek** [OpenTTD/OpenTTD#6598]
**_james1101_ opened the ticket and wrote:**

> To Reproduce this Crash:
> Start OTTD server.
> Remember it's ip address.
> Start OTTD client.
> Client: Connect to server.
> Client: Execute 'join 255'.
> \-> Client joins spectators.
> Client: Execute 'join N', where N is greater than 15 and less than 255.
> \-> Client throws: 'Error: Company does not exist.'
> Client: Execute 'connect <server-ip># N', where N is greater than 15 and less than 255.
> \-> Client: Connect should be successful.
> Client: Open company list.
> \-> Client crashes. Crash files are attached.
> 
> Client version 1.7.1 on Windows 7 Ultimate; Server version 1.7.1 on Windows 7 Ultimate
> 
> Suggested Fix: Make the company id part of the connect command's argument follow the same rules as the join command's company id argument, ie must be between 0 and max_companies, or 255 (spectator).

**Attachments**
- [crash.png](https://bugs.openttd.org/task/6598/getfile/10764/crash.png) (211.85 KiB)
- [crash.log](https://bugs.openttd.org/task/6598/getfile/10765/crash.log) (8.83 KiB)
- [crash.dmp](https://bugs.openttd.org/task/6598/getfile/10766/crash.dmp) (12978.97 KiB)

**Reported version**: _1.7.1_
**Operating system**: _All_

---

**Comment by DorpsGek** [comment]
**_Wolf01_ wrote:**

> I can't reproduce it on Win 10.
> Tried with local server both with IPv4 and IPv6.
> Tried logging in as a company 1 and spectator.
> Tried connecting from titlescreen.
> 
> The command already checks for join_as > MAX_COMPANIES:
```
 if (join_as != COMPANY_SPECTATOR) {
 	if (join_as > MAX_COMPANIES) return false;
 	join_as--;
 }
```

---
###### _This comment was imported from FlySpray: https://bugs.openttd.org/task/6598#comment14523_

**Comment by DorpsGek** [comment]
**__dp__ wrote:**

> Ok, I didn't get any crashes, but I think it's close enough. Thing is connect cmd sometimes leaves the game in a weird half-disconnected state. It's not very noticable at first but it's clearly not a valid state. E.g. you can build rails but can't leave company or use join command. Or you can bring up mp lobby while still being in game.
> There are several ways to get in that state:
> 1) connect ::1# 254   (or anything 16-254)
> 2) connect ::1# 1 and cancelling password prompt (company 1 has to be passworded)
> 3) connect ::1:0
> 
> I think I got all such cases in my patch but can't be 100% sure.
> Also rearranged console output a bit so it doesn't say "connecting ..." before failing on arg parsing.
> And added check for negative company #  to prevent possible bugs with it.

**Attachments**
- [6598-connect-cmd-fix.patch](https://bugs.openttd.org/task/6598/getfile/10783/6598-connect-cmd-fix.patch) (1.54 KiB)

---
###### _This comment was imported from FlySpray: https://bugs.openttd.org/task/6598#comment14526_

**Comment by TrueBrain** [comment]
Patch contains many things I am unsure are part of the solution. But it is a good start for solving this issue :)

**Comment by andythenorth** [comment]
Thanks for this.  Valid issue, but there's been no activity on this for some time, and as it stands, it doesn't look likely that it will go any further.  I'm closing it as we try to keep the issue count low for OpenTTD, it helps us focus on things that are important and fun.  Feel free to discuss in irc or request re-opening if you disagree.  Thanks for contributing!

**Comment by nikolas** [comment]
Here's a [crash.log](https://github.com/OpenTTD/OpenTTD/files/3470101/crash.log) triggered by a related scenario:
* Enable an AI player to start in the game immediately. I used AdmiralAI.
* Host a game, and put a password on it
* From a second client, join the game, and join that company. (`connect ::1#1` from the console.)
* In the console of the second client, do: `connect ::1#1` (effectively trying to join the same company). Don't type in the server password yet.
* In this state, where the server password is being prompted, you can crash the second client in any number of ways. The first ones I've found: Open up the Company window, and try and change the company name, or president's name. Cancelling the password dialog box causes a crash as well.

Additionally, writing the crash savegame fails here (resolved by #7684):

> Writing crash savegame...
> Error: Assertion failed at line 286 of /home/nik/src/OpenTTD/src/ai/ai_core.cpp: c != nullptr && c->ai_instance != nullptr
> Aborted

**Comment by Foxar** [comment]
Since this is marked as good first issue and I'm a newbie, I'd love to try and help out with this, could anyone point me in the right direction? :)
